cachito/workers/tasks/gomod.py

```diff
@@ -116,14 +121,31 @@ def fetch_gomod_source(request_id, dep_replacements=None, package_configs=None):
         update_request_with_package(request_id, gomod["module"], env_vars, package_subpath=subpath)
         update_request_with_deps(request_id, gomod["module"], gomod["module_deps"])
 
+        packages_json_data["packages"].append({
+            "dependencies": gomod["module_deps"],
+            "name": gomod["module"]["name"],
+            "type": gomod["module"]["type"],
+            "version": gomod["module"]["version"],
+            "path": subpath,
+        })
+
         # add package deps
         for package in gomod["packages"]:
-            package_subpath = _package_subpath(
-                gomod["module"]["name"], package["pkg"]["name"], subpath
-            )
-            update_request_with_package(request_id, package["pkg"], package_subpath=package_subpath)
+            pkg_info = package["pkg"]
+            package_subpath = _package_subpath(gomod["module"]["name"], pkg_info["name"], subpath)
+            update_request_with_package(request_id, pkg_info, package_subpath=package_subpath)
             if package.get("pkg_deps"):
-                update_request_with_deps(request_id, package["pkg"], package["pkg_deps"])
+                update_request_with_deps(request_id, pkg_info, package["pkg_deps"])
+
+            packages_json_data["packages"].append({
+                "dependencies": package["pkg_deps"],
+                "name": pkg_info["name"],
+                "type": pkg_info["type"],
+                "version": pkg_info["version"],
+                "path": package_subpath,
+            })
+
+    bundle_dir.gomod_packages_data.write_text(json.dumps(packages_json_data), encoding="utf-8")
```

Comment:
I realize this is a draft, just gonna comment this so that I don't forget :D

It might be worth it to check for duplicate packages. I think packages should be unique by `name, type, version`. If there are duplicates IMO we should raise an error.

Previously, it was technically possible to have two packages with the same `name, type, version` as long as they had the same `path`. If those packages somehow had different lists of dependencies, cachito would merge the lists into one. If you had two identical packages with different paths, Cachito would raise an error.

But if two packages have the same `name, type, version, path`, then they are the same package and we should never process them more than once during a cachito request. So if two packages have the same `name, type, version`, that should always be an error regardless of the `path`.

It would probably be good to check for duplicates in each of the `{gomod,npm,pip,yarn}_package.json` files as well as the final `{request_id}-packages.json`

Reply:
@chmeliik Is `resolve_*` function the good place to make this check?

Reply:
I think `fetch_*_source` tasks would be the right place, because they combine the results of multiple `resolve_*` calls. We should be checking that packages are still unique even after combining all the resolve calls.